feat: Add failed node info to exit handler

simster7 opened the pull request:
Closes #1994 

Adds failed node information to the exit handler as a JSON list under the variable `workflow.failures`.

Example:
```yaml
# Example on continue-on-fail, where the workflow will continue running even
# some steps fail.
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: continue-on-fail-
spec:
  entrypoint: workflow-ignore
  onExit: send-email
  parallelism: 1
  templates:
  - name: workflow-ignore
    steps:
    - - name: A
        template: intentional-fail
        continueOn:
          failed: true
    - - name: B
        template: whalesay
      - name: C
        template: intentional-fail
        continueOn:
          failed: true
    - - name: D
        template: whalesay

  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["hello world"]

  - name: intentional-fail
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo intentional failure; exit 1"]

  - name: send-email
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo send e-mail: {{workflow.name}} {{workflow.status}}. Failed steps {{workflow.failures}}"]
```

`workflow.failures` contents:
```json
[
  {
    "name": "C",
    "message": "failed with exit code 1",
    "templateName": "intentional-fail",
    "finishedAt": "2020-02-04 18:07:48 +0000 UTC",
    "phase": "Failed"
  },
  {
    "name": "A",
    "message": "failed with exit code 1",
    "templateName": "intentional-fail",
    "finishedAt": "2020-02-04 18:07:37 +0000 UTC",
    "phase": "Failed"
  }
]
```

simster7 commented:
I'm thinking it would be better to name the parameter `workflow.failures`. Or perhaps something else? Opinions?

alexec commented on workflow/controller/operator.go:
```diff
@@ -262,6 +262,33 @@ func (woc *wfOperationCtx) operate() {
 			woc.globalParams[common.GlobalVarWorkflowStatus] = string(workflowStatus)
 		}
+
+		var failures []interface{}
```
why not create a named struct?

simster7 replied:
Since this is a subset of `NodeStatus`, I did not want to create an explicit new type.

alexec commented on workflow/controller/operator.go:
```diff
@@ -262,6 +262,33 @@ func (woc *wfOperationCtx) operate() {
+					Name:         node.DisplayName,
+					Message:      node.Message,
+					TemplateName: node.TemplateName,
+					FinishedAt:   node.FinishedAt.String(),
```
what timezone is this in?

simster7 replied:
It is the same information from `NodeStatus`, so whichever timezone that is in

alexec commented on workflow/controller/operator.go:
```diff
@@ -262,6 +262,33 @@ func (woc *wfOperationCtx) operate() {
+					Name         string `json:"name"`
+					Message      string `json:"message"`
+					TemplateName string `json:"templateName"`
+					FinishedAt   string `json:"finishedAt"`
```
these can be typed?

simster7 replied:
Don't think that's super necessary, since this is an ephemeral struct

alexec replied:
change `finishedAt` to `time.Time` and put into a documented struct

samath117 commented:
This doesn't give easy access to the logs of the failed step(s). If you don't want to include all of those, could you at least give their pod name(s)?

simster7 commented:
> This doesn't give easy access to the logs of the failed step(s). If you don't want to include all of those, could you at least give their pod name(s)?

Great idea... will add

simster7 commented:
> Great idea... will add

Added

alexec commented on workflow/common/common.go:
```diff
@@ -118,6 +118,8 @@ const (
 	// GlobalVarWorkflowPriority is the workflow variable referencing the workflow's priority field
 	GlobalVarWorkflowPriority = "workflow.priority"
+	// GobalVarWorkflowFailedNodes is a global workflow variable referencing the workflow's failed nodes
+	GobalVarWorkflowFailedNodes = "workflow.failures"
```
`GobalVarWorkflowFailedNodes` -> `GlobalVarWorkflowFailedNodes`

alexec replied:
can you document that it is an array please?

alexec commented on workflow/controller/operator.go:
```diff
@@ -262,6 +262,35 @@ func (woc *wfOperationCtx) operate() {
+		for _, node := range woc.wf.Status.Nodes {
+			if node.Phase == wfv1.NodeFailed || node.Phase == wfv1.NodeError {
+				failures = append(failures, struct {
+					Name         string `json:"name"`
```
can you please change from `Name` to `DisplayName`

simster7 commented:
@alexec Addressed your concerns! However, I decided not to include a full Slack example. My reasoning: we shouldn't provide documentation about how to send slack messages, as this could change any time and should be looked at from Slack's own docs.

However, I did include an example of how to manipulate the JSON map in bash using `jq` which is arguably the difficult part of sending the slack message and demonstrates explicitly that `{{workflow.failures}}` is indeed a JSON
